c_src/ex_dtls/native.c

```diff
@@ -92,90 +128,136 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
       // function.
-      gen_packets = (UnifexPayload *)unifex_payload_alloc(
-          env, UNIFEX_PAYLOAD_BINARY, dyn_buff->data_size);
-      memcpy(gen_packets->data, dyn_buff->data, dyn_buff->data_size);
-      gen_packets->size = (unsigned int)dyn_buff->data_size;
-      res_term = do_handshake_result_ok(env, state, gen_packets);
-      dyn_buff_free(dyn_buff);
-      unifex_payload_release(gen_packets);
-      return res_term;
-    case SSL_ERROR_WANT_WRITE:
-      DEBUG("SSL WANT WRITE");
-      break;
-    case SSL_ERROR_NONE:
-      /*
-        This will be reached only when handshake succeeds because SSL_ERROR_NONE
-        is returned only when res (passed to SSL_get_error) is greater than 0
-        and res greater than zero means handshake succeeded.
-      */
-      DEBUG("Handshake finished successfully");
-
-      KeyingMaterial *keying_material = export_keying_material(state->ssl);
-      if (keying_material == NULL) {
-        DEBUG("Cannot export keying material");
-        return unifex_raise(env, "Handshake failed: cannot export keying material");
-      }
-
-      int len = keying_material->len;
-      UnifexPayload *client_keying_material =
-          (UnifexPayload *)unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, len);
-      memcpy(client_keying_material->data, keying_material->client, len);
-      client_keying_material->size = len;
-
-      UnifexPayload *server_keying_material =
-          (UnifexPayload *)unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, len);
-      memcpy(server_keying_material->data, keying_material->server, len);
-      server_keying_material->size = len;
-
+      DEBUG("WBIO: read: %d bytes", read_bytes);
       UnifexPayload *gen_packets = (UnifexPayload *)unifex_payload_alloc(
-          env, UNIFEX_PAYLOAD_BINARY, dyn_buff->data_size);
-      memcpy(gen_packets->data, dyn_buff->data, dyn_buff->data_size);
-      gen_packets->size = (unsigned int)dyn_buff->data_size;
-      res_term = do_handshake_result_finished(
-          env, state, client_keying_material, server_keying_material,
+          state->env, UNIFEX_PAYLOAD_BINARY, pending_data_len);
+      memcpy(gen_packets->data, pending_data, pending_data_len);
+      gen_packets->size = (unsigned int)pending_data_len;
+      state->hsk_finished = 1;
+      UNIFEX_TERM res_term = process_result_hsk_finished(
+          state->env, state, client_keying_material, server_keying_material,
           keying_material->protection_profile, gen_packets);
-      dyn_buff_free(dyn_buff);
+      free(pending_data);
       unifex_payload_release(gen_packets);
       unifex_payload_release(client_keying_material);
       unifex_payload_release(server_keying_material);
       return res_term;
-    default:
-      DEBUG("SSL ERROR: %d", res);
-      return unifex_raise(env, "Handshake failed: SSL error");
     }
+  } else {
+    // we have no pending data to send - send empty binary
```

Comment:
Can't we just send nothing?

Reply:
I think that after refactoring it looks quite good. We got rid of this else [here](https://github.com/membraneframework/ex_dtls/blob/dtls-alert/c_src/ex_dtls/native.c#L188-L196).
Now we always allocate payload for pending data. When pending data is > 0 we copy it to payload. In other case we just return empty payload. Sending nothing would require additional return type in `native.spec.exs` e.g. `handshake_finished_with_packets`.